File: Source/Extensions/Tessa.Extensions.Default.Client/Forums/TopicsUIExtension.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared.Workflow.KrPermissions;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Forums;
using Tessa.Forums.ForumCached;
using Tessa.Forums.Notifications;
using Tessa.Platform;
using Tessa.Platform.Collections;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Controls.Forums;
using Tessa.UI.Controls.Forums.Controls;
using Tessa.UI.Windows;
using Tessa.Views.Parser.SyntaxTree.Workplace;

namespace Tessa.Extensions.Default.Client.Forums
{
    /// <summary>
    /// Класс для управления вкладкой обсуждения
    /// </summary>
    public sealed class TopicsUIExtension : CardUIExtension
    {
        #region Constructors

        public TopicsUIExtension(
            IForumDialogManager forumDialogManager,
            IDocumentTabManager documentTabManager,
            IWorkplaceInterpreter interpreter)
        {
            this.documentTabManager = documentTabManager;
            this.interpreter = interpreter;
            this.forumDialogManager = forumDialogManager;
        }

        #endregion

        #region Base overrides

        public override async Task Initialized(ICardUIExtensionContext context)
        {
            if (!context.ValidationResult.IsSuccessful())
            {
                return;
            }

            var cardID = context.Card.ID;
            var uiContext = context.UIContext;
            var forms = context.Model.Forms;

            // проверить что карточка входит в типое решине и на ней включено "использовать систему форумов"
            foreach (ForumControlViewModel forumControl in GetAllForumControls(context))
            {
                // обновляем всю модель при выборе вкладки
                if (forumControl.Content is ForumLicenseNotExistViewModel)
                {
                    return;
                }

                if (forumControl.Block.Form.IsCollapsed
                    || !forms.Contains(forumControl.Block.Form))
                {
                    continue;
                }

                if (context.Model.InSpecialMode())
                {
                    forumControl.EnableSpecialMode();
                }

                forumControl.OpenParticipants(
                    (openParticipantsContext, modifyParamAction) => this.OpenParticipantsView(cardID, openParticipantsContext, modifyParamAction));
                forumControl.CheckAddTopicPermission(
                    ct => this.CheckAddTopicPermissionsAsync(forumControl, uiContext, ct));
                forumControl.CheckSuperModeratorPermissionAsync(
                    ct => CheckSuperModeratorPermissionsAsync(uiContext, ct));

                // Если в инфо есть topicID значит мы пришли из OpenTopicOnDoubleClickExtension. и Это ID выбронного топика
                if (context.Model.Card.Info.TryGetValue(ForumHelper.TopicIDKey, out object t) && t is Guid topicID &&
                    context.Model.Card.Info.TryGetValue(ForumHelper.TopicTypeIDKey, out object type) && type is Guid typeID)
                {
                    if (forumControl.ForumViewModel.TopicTypeID == typeID)
                    {
                        IForumData data = forumControl.ForumViewModel.ForumControlDependencies.ForumClientCachedDataManager.GetForumData();
                        await forumControl.SelectTopicAsync(
                            topicID,
                            null,
                            data.ReadTopicStatusList.TryFirst(
                                top => top.TopicID == topicID, out UserStatusModel userStat)
                                ? userStat.LastReadMessageTime
                                : null,
                            cancellationToken: context.CancellationToken);
                        context.Model.Card.Info.Remove(ForumHelper.TopicIDKey);

                        var mainForm = context.Model.MainFormWithTabs;
                        mainForm.SelectedTab = forumControl.Block.Form;
                    }
                }
            }
        }

        #endregion

        #region Fields

        private readonly IForumDialogManager forumDialogManager;
        private readonly IDocumentTabManager documentTabManager;
        private readonly IWorkplaceInterpreter interpreter;

        #endregion

        #region Private

        /// <summary>
        /// Возвращает все топики в алиаcе которого есть название Forum
        /// </summary>
        private static List<ForumControlViewModel> GetAllForumControls(ICardUIExtensionContext context) =>
            context.Model.ControlBag.OfType<ForumControlViewModel>().ToList();


        private async Task CheckAddTopicPermissionsAsync(ForumControlViewModel forumControlViewModel, IUIContext uiContext, CancellationToken cancellationToken = default)
        {
            await OpenMarkedCardAsync(
                uiContext,
                KrPermissionsHelper.CalculateAddTopicPermissions,
                null, //Не требуем подтверждения действия, если не было изменений
                cardIsNew => cardIsNew
                    ? TessaDialog.Confirm("$KrTiles_EditModeConfirmation") ? true : null
                    : TessaDialog.ConfirmWithCancel("$KrTiles_EditModeConfirmation"),
                () => this.AddTopicShowDialogAsync(forumControlViewModel, cancellationToken),
                cancellationToken: cancellationToken);
        }

        private static async Task CheckSuperModeratorPermissionsAsync(IUIContext uiContext, CancellationToken cancellationToken = default)
        {
            await OpenMarkedCardAsync(
                uiContext,
                KrPermissionsHelper.CalculateSuperModeratorPermissions,
                null, //Не требуем подтверждения действия, если не было изменений
                cardIsNew => cardIsNew
                    ? TessaDialog.Confirm("$KrTiles_EditModeConfirmation") ? true : null
                    : TessaDialog.ConfirmWithCancel("$KrTiles_EditModeConfirmation"),
                async () => await SuperModeratorPermissionsMessageAsync(),
                cancellationToken: cancellationToken);
        }

        private async Task<bool> AddTopicShowDialogAsync(ForumControlViewModel forumControlViewModel, CancellationToken cancellationToken = default)
        {
            var card = UIContext.Current.CardEditor?.CardModel?.Card;
            if (card is null)
            {
                return false;
            }

            var token = KrToken.TryGet(card.Info);
            if (token?.HasPermission(KrPermissionFlagDescriptors.AddTopics) == true)
            {
                await this.forumDialogManager.AddTopicShowDialogAsync(card.ID, model => forumControlViewModel.ForumViewModel.ModifyAddingTopic(model), cancellationToken);
                return true;
            }

            TessaDialog.ShowError("$Forum_Permission_NoPermissionToAddTopic");
            return false;
        }

        private static async Task<bool> SuperModeratorPermissionsMessageAsync()
        {
            var card = UIContext.Current.CardEditor?.CardModel?.Card;
            if (card is null)
            {
                return false;
            }

            var token = KrToken.TryGet(card.Info);
            if (token?.HasPermission(KrPermissionFlagDescriptors.SuperModeratorMode) == true)
            {
                TessaDialog.ShowMessage("$Forum_Permission_SuperModeratorModeOn");
                return false;
            }

            TessaDialog.ShowError("$Forum_Permission_NoRequiredPermissions");
            return false;
        }

        private static async Task OpenMarkedCardAsync(IUIContext context,
            string mark,
            Func<bool> proceedConfirmation,
            Func<bool, bool?> proceedAndSaveCardConfirmation,
            Func<Task<bool>> continuationOnSuccessFunc = null,
            Dictionary<string, object> getInfo = null,
            CancellationToken cancellationToken = default)
        {
            ICardEditorModel editor = context.CardEditor;
            ICardModel model;

            if (editor is null || editor.OperationInProgress || (model = editor.CardModel) is null)
            {
                return;
            }

            bool cardIsNew = model.Card.StoreMode == CardStoreMode.Insert;
            bool hasChanges = cardIsNew || await model.HasChangesAsync(cancellationToken: cancellationToken);
            bool? saveCardBeforeOpening;

            if (hasChanges && proceedAndSaveCardConfirmation is not null)
            {
                saveCardBeforeOpening = proceedAndSaveCardConfirmation(cardIsNew);
            }
            //Если не указана функция подтверждения с вариантом отмены - сохраняем карточку
            //если есть подтверждение основного действия
            else if (proceedConfirmation is not null && hasChanges)
            {
                saveCardBeforeOpening = proceedConfirmation() ? true : null;
            }
            //Если в карточке не было изменений - не вызываем сохранения
            else if (proceedConfirmation is not null)
            {
                saveCardBeforeOpening = proceedConfirmation() ? false : null;
            }
            //Если не указана функция подтверждения и нет изменений - вызываем основное действие
            //без подтверждения и сохранения
            else
            {
                saveCardBeforeOpening = false;
            }

            getInfo ??= new Dictionary<string, object>(StringComparer.Ordinal);

            if (!string.IsNullOrWhiteSpace(mark))
            {
                getInfo[mark] = BooleanBoxes.True;
            }

            if (!saveCardBeforeOpening.HasValue)
            {
                return;
            }

            if (saveCardBeforeOpening.Value)
            {
                KrToken token = KrToken.TryGet(editor.Info);
                KrToken.Remove(editor.Info);

                bool res = await editor.SaveCardAsync(
                    context,
                    new Dictionary<string, object>
                    {
                        { KrPermissionsHelper.SaveWithPermissionsCalcFlag, true }
                    },
                    new CardSavingRequest(CardSavingMode.KeepPreviousCard),
                    cancellationToken);
                if (!res)
                {
                    return;
                }

                token?.Set(getInfo);
            }

            Guid cardID = model.Card.ID;
            CardType cardType = model.CardType;

            bool sendTaskSucceeded = await editor.OpenCardAsync(
                cardID,
                cardType.ID,
                cardType.Name,
                context,
                getInfo,
                cancellationToken: cancellationToken);

            if (sendTaskSucceeded)
            {
                editor.IsUpdatedServer = true;
            }
            else if (cardIsNew || saveCardBeforeOpening.Value)
            {
                // если карточка новая или была сохранена, а также не удалось выполнить mark-действие при открытии,
                // то у нас будет "висеть" карточка с некорректной версией;
                // её надо обновить, на этот раз без mark'и

                await editor.OpenCardAsync(
                    cardID,
                    cardType.ID,
                    cardType.Name,
                    context,
                    cancellationToken: cancellationToken);
            }

            if (continuationOnSuccessFunc is null)
            {
                return;
            }

            await using (UIContext.Create(context))
            {
                await continuationOnSuccessFunc();
            }
        }

        private async ValueTask OpenParticipantsView(
            Guid cardID,
            OpenParticipantsContext openParticipantsContext,
            Func<OpenParticipantsContext, ValueTask> modifyParamAction = null)
        {
            if (modifyParamAction is not null)
            {
                await modifyParamAction(openParticipantsContext);
            }

            await TopicParticipantsWorkplaceTab.OpenParticipantsViewTabAsync(
                this.interpreter,
                this.documentTabManager,
                cardID,
                openParticipantsContext);
        }

        #endregion
    }
}

File: WebClient SDK/src/default/forums/topicsUIExtension.ts
import {
  CardSavingMode,
  CardSavingRequest,
  CardUIExtension,
  ICardModel,
  ICardUIExtensionContext
} from 'tessa/ui/cards';
import { ForumActionParameters, ForumDialogManager, ForumViewModel } from 'tessa/ui/cards/controls';
import {
  IUIContext,
  showConfirm,
  showConfirmWithCancel,
  showError,
  showMessage,
  tryGetFromInfo,
  UIContext
} from 'tessa/ui';
import { CardStoreMode } from 'tessa/cards';
import { createTypedField, DotNetType } from 'tessa/platform';
import { AddTopics, KrToken, SuperModeratorMode } from 'tessa/workflow';
import { IStorage } from 'tessa/platform/storage';
import { equalsCriteriaOperator, ViewParameterMetadata } from 'tessa/views/metadata';
import { ForumHelper } from 'tessa/forums';
import { RequestParameterBuilder } from 'tessa/views';
import { showView } from 'tessa/ui/uiHost';
import { DefaultFormMainViewModel } from 'tessa/ui/cards/forms';

export class TopicsUIExtension extends CardUIExtension {
  //#region fields

  private _dispose: (() => void) | null = null;

  //#endregion

  //#region CardUIExtension

  public async initialized(context: ICardUIExtensionContext): Promise<void> {
    if (!context.validationResult.isSuccessful) {
      return;
    }

    const allForumControls = context.model.controlsBag.filter(
      x => x instanceof ForumViewModel
    ) as ForumViewModel[];
    for (const forumControl of allForumControls) {
      if (!forumControl.isLicenseValid) {
        continue;
      }

      if (forumControl.block.form.isCollapsed) {
        continue;
      }

      if (context.model.inSpecialMode) {
        forumControl.isAddTopicEnabled = false;
        forumControl.isEnabledForumEmptyContextMenu = true;
      }

      forumControl.openParticipantsAction = this.getOpenParticipantsAction(context);

      forumControl.checkAddTopicPermissionAction = this.getCheckAddTopicPermissionsAction(
        context,
        forumControl
      );

      forumControl.checkSuperModeratorPermissionAction =
        this.getCheckSuperModeratorPermissionAction(context);

      const topicId = tryGetFromInfo<guid>(context.model.card.info, ForumHelper.TopicIDKey);
      const topicTypeId = tryGetFromInfo<guid>(context.model.card.info, ForumHelper.TopicTypeIDKey);
      if (topicId && topicTypeId && forumControl.topicTypeId === topicTypeId) {
        await forumControl.showTopic(topicId);
        delete context.model.card.info[ForumHelper.TopicIDKey];
        delete context.model.card.info[ForumHelper.TopicTypeIDKey];
        const mainFrom = context.model.mainForm as DefaultFormMainViewModel;
        mainFrom.selectedTab = forumControl.block.form;
        return;
      }
    }
  }

  public finalized(): void {
    if (this._dispose) {
      this._dispose();
      this._dispose = null;
    }
  }

  //#endregion

  //#region actions

  private getOpenParticipantsAction(context: ICardUIExtensionContext) {
    const openParticipantsAction: (
      params: ForumActionParameters,
      modifyOpenParticipantsAction: (params: ForumActionParameters) => void
    ) => Promise<void> = async (params, modifyOpenParticipants): Promise<void> => {
      modifyOpenParticipants && modifyOpenParticipants(params);

      const paramTopicMeta = new ViewParameterMetadata();
      paramTopicMeta.alias = 'TopicID';

      const paramTopicId = new RequestParameterBuilder()
        .withMetadata(paramTopicMeta)
        .addCriteria(equalsCriteriaOperator(), 'topicID', params.topicID)
        .asRequestParameter();

      const paramCardMeta = new ViewParameterMetadata();
      paramCardMeta.alias = 'CardID';
      const paramCardId = new RequestParameterBuilder()
        .withMetadata(paramCardMeta)
        .addCriteria(equalsCriteriaOperator(), 'cardID', context.card.id)
        .asRequestParameter();

      const paramParticipantMeta = new ViewParameterMetadata();
      paramParticipantMeta.alias = 'ParticipantTypeID';
      const paramParticipantType = new RequestParameterBuilder()
        .withMetadata(paramParticipantMeta)
        .addCriteria(equalsCriteriaOperator(), 'participantType', params.participantsTypeID)
        .asRequestParameter();

      await showView({
        viewAlias: 'TopicParticipants',
        displayValue: '$Workplaces_User_TopicParticipants',
        parameters: [paramTopicId, paramCardId, paramParticipantType],
        treeVisible: false
      });
    };

    return openParticipantsAction;
  }

  private getCheckAddTopicPermissionsAction(
    context: ICardUIExtensionContext,
    control: ForumViewModel
  ) {
    return async (): Promise<void> => {
      await TopicsUIExtension.openMarkedCard(
        context.uiContext,
        'kr_calculate_addtopic_permissions',
        null, // Не требуем подтверждения действия, если не было изменений
        async cardIsNew =>
          cardIsNew
            ? (await showConfirm('$KrTiles_EditModeConfirmation'))
              ? true
              : null
            : await showConfirmWithCancel('$KrTiles_EditModeConfirmation'),
        async () => await this.addTopicShowDialog(context, control)
      );
    };
  }

  private getCheckSuperModeratorPermissionAction(context: ICardUIExtensionContext) {
    return async (): Promise<void> => {
      await TopicsUIExtension.openMarkedCard(
        context.uiContext,
        'kr_calculate_supermoderator_permissions',
        null, // Не требуем подтверждения действия, если не было изменений
        async cardIsNew =>
          cardIsNew
            ? (await showConfirm('$KrTiles_EditModeConfirmation'))
              ? true
              : null
            : await showConfirmWithCancel('$KrTiles_EditModeConfirmation'),
        async () => await this.superModeratorPermissionsMessage()
      );
    };
  }

  private addTopicShowDialog = async (
    context: ICardUIExtensionContext,
    control: ForumViewModel
  ) => {
    const card = UIContext.current.cardEditor?.cardModel?.card;
    if (!card) {
      return false;
    }

    const token = KrToken.tryGet(card.info);
    if (token?.hasPermission(AddTopics) == true) {
      await ForumDialogManager.instance.addTopicShowDialog(context.card.id, model =>
        control.modifyAddingTopicAction(model)
      );
      return true;
    } else {
      await showError('$Forum_Permission_NoPermissionToAddTopic');
      return false;
    }
  };

  private superModeratorPermissionsMessage = async () => {
    const card = UIContext.current.cardEditor?.cardModel?.card;
    if (!card) {
      return false;
    }

    const token = KrToken.tryGet(card.info);
    if (token?.hasPermission(SuperModeratorMode) === true) {
      await showMessage('$Forum_Permission_SuperModeratorModeOn');
      return false;
    }

    await showError('$Forum_Permission_NoRequiredPermissions');
    return false;
  };

  private static async openMarkedCard(
    context: IUIContext,
    mark: string | null,
    proceedConfirmation: (() => Promise<boolean>) | null,
    proceedAndSaveCardConfirmation: ((cardIsNew: boolean) => Promise<boolean | null>) | null,
    continuationOnSuccessFunc: (() => Promise<boolean>) | null = null,
    getInfo: IStorage | null = null
  ): Promise<void> {
    const editor = context.cardEditor;
    let model: ICardModel;

    if (!editor || editor.operationInProgress || !(model = editor.cardModel!)) {
      return;
    }

    const cardIsNew = model.card.storeMode === CardStoreMode.Insert;
    const hasChanges = cardIsNew || (await model.hasChanges());
    let saveCardBeforeOpening: boolean | null;

    if (hasChanges && proceedAndSaveCardConfirmation) {
      saveCardBeforeOpening = await proceedAndSaveCardConfirmation(cardIsNew);
      // Если не указана функция подтверждения с вариантом отмены - сохраняем карточку
      // если есть подтверждение основного действия
    } else if (hasChanges && proceedConfirmation) {
      saveCardBeforeOpening = (await proceedConfirmation()) ? true : null;
      // Если в карточке не было изменений - не вызываем сохранения
    } else if (proceedConfirmation) {
      saveCardBeforeOpening = (await proceedConfirmation()) ? false : null;
      // Если не указана функция подтверждения и нет изменений - вызываем основное действие
      // без подтверждения и сохранения
    } else {
      saveCardBeforeOpening = false;
    }

    if (saveCardBeforeOpening === null) {
      return;
    }

    if (!getInfo) {
      getInfo = {};
    }

    if (mark) {
      getInfo[mark] = createTypedField(true, DotNetType.Boolean);
    }

    if (saveCardBeforeOpening) {
      const token = KrToken.tryGet(editor.info);
      KrToken.remove(editor.info);

      const saveSuccess = await editor.saveCard(
        context,
        {
          '.SaveWithPermissionsCalc': createTypedField(true, DotNetType.Boolean)
        },
        new CardSavingRequest(CardSavingMode.KeepPreviousCard)
      );

      if (!saveSuccess) {
        return;
      }

      if (token) {
        token.setInfo(getInfo);
      }
    }

    const cardId = model.card.id;
    const cardType = model.cardType;

    const sendTaskSucceeded = await editor.openCard({
      cardId,
      cardTypeId: cardType.id!,
      cardTypeName: cardType.name!,
      context,
      info: getInfo
    });

    if (sendTaskSucceeded) {
      editor.isUpdatedServer = true;
    } else if (cardIsNew || saveCardBeforeOpening) {
      // если карточка новая или была сохранена, а также не удалось выполнить mark-действие при открытии,
      // то у нас будет "висеть" карточка с некорректной версией;
      // её надо обновить, на этот раз без mark'и

      await editor.openCard({
        cardId,
        cardTypeId: cardType.id!,
        cardTypeName: cardType.name!,
        context
      });
    }

    if (!continuationOnSuccessFunc) {
      return;
    }

    const contextInstance = UIContext.create(context);
    try {
      await continuationOnSuccessFunc();
    } finally {
      contextInstance.dispose();
    }
  }

  //#endregion
}
